ddht/v5_1/alexandria/client.py

```diff
@@ -110,6 +131,93 @@ async def _request(
+
+    @asynccontextmanager
+    async def subscribe_request(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        request: AlexandriaMessage[Any],
+        response_message_type: Type[TAlexandriaMessage],
+        request_id: Optional[bytes],
+    ) -> AsyncIterator[trio.abc.ReceiveChannel[InboundMessage[TAlexandriaMessage]]]:
+        if request_id is None:
+            request_id = self.network.client.request_tracker.get_free_request_id(
+                node_id
+            )
+
+        self.logger.debug(
+            "Sending request: %s with request id %s", request, request_id.hex(),
+        )
+
+        send_channel, receive_channel = trio.open_memory_channel[
+            InboundMessage[TAlexandriaMessage]
+        ](256)
+
+        with self.request_tracker.reserve_request_id(node_id, request_id):
+            async with trio.open_nursery() as nursery:
+                nursery.start_soon(
+                    functools.partial(
+                        self._manage_request_response,
+                        node_id,
+                        endpoint,
+                        request,
+                        response_message_type,
+                        send_channel,
+                        request_id,
+                    )
+                )
+                try:
+                    async with receive_channel:
+                        yield receive_channel
+                finally:
+                    nursery.cancel_scope.cancel()
+
+    async def _manage_request_response(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        request: AlexandriaMessage[Any],
+        response_message_type: Type[AlexandriaMessage[Any]],
+        send_channel: trio.abc.SendChannel[InboundMessage[AlexandriaMessage[Any]]],
+        request_id: bytes,
+    ) -> None:
+        with trio.move_on_after(REQUEST_RESPONSE_TIMEOUT) as scope:
+            subscription_ctx = self.subscription_manager.subscribe(
+                response_message_type, endpoint, node_id,
+            )
+            async with subscription_ctx as subscription:
+                self.logger.debug(
+                    "Sending request with request id %s", request_id.hex(),
+                )
+                # Send the request
+                await self._send_request(
+                    node_id, endpoint, request, request_id=request_id
+                )
+
+                # Wait for the response
+                async with send_channel:
+                    async for response in subscription:
+                        if response.request_id != request_id:
+                            continue
+                        else:
+                            await send_channel.send(response)
+        if scope.cancelled_caught:
+            self.logger.warning(
+                "Abandoned request response monitor: request=%s message_type=%s",
+                request,
+                response_message_type,
+            )
```

Comment:
If I'm reading this right, won't this warning always be emitted? There's no `break` in the inner loop, it will keep trying to shuttle messages from one channel to another until it times out.

Reply:
Yeah, I think you're correct (and I'm seeing this in the base protocol as well)

Reply:
Ok, no, not actually true.  So this function is run within a nursery.  When the outer nursery is cancelled, it doesn't trigger this because trio is smart and the cancellations are scoped such that they don't leak between cancel scopes.